File: 4 Median of Two Sorted Arrays/4 LL.cpp
#include <vector>
#include <iostream>
using namespace std;

/*
https://leetcode.com/problems/median-of-two-sorted-arrays/

There are two sorted arrays nums1 and nums2 of size m and n respectively.
Find the median of the two sorted arrays.
The overall run time complexity should be O(log (m+n)).
*/

class Solution
{
private:
	int findKth(vector<int>::iterator a, int lena, vector<int>::iterator b, int lenb, int k)
	{
		if (lena == 0) return b[k - 1];
		if (lenb == 0) return a[k - 1];
		if (k == 1) return min(a[0], b[0]);
		int posa = min(k >> 1, lena - 1);
		int posb = min(max(0, k - posa - 2), lenb - 1);
		posa = k - posb - 2;
		if (a[posa] <= b[posb])
		{
			if ((posa == lena - 1) || (b[posb] <= a[posa + 1]))
			{
				return b[posb];
			}
			return findKth(a + posa + 1, lena - posa - 1, b, lenb, k - posa - 1);
		}
		else
		{
			if ((posb == lenb - 1) || (a[posa] <= b[posb + 1]))
			{
				return a[posa];
			}
			return findKth(a, lena, b + posb + 1, lenb - posb - 1, k - posb - 1);
		}
	}
public:
	double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2)
	{
		double answer = 0;
		int total = nums1.size() + nums2.size();
		if (total != 0)
		{
			if ((total & 0x1) != 0)
			{
				answer += findKth(nums1.begin(), nums1.size(), nums2.begin(), nums2.size(), (total + 1) >> 1);
			}
			else
			{
				answer += findKth(nums1.begin(), nums1.size(), nums2.begin(), nums2.size(), total >> 1);
				answer += findKth(nums1.begin(), nums1.size(), nums2.begin(), nums2.size(), (total >> 1) + 1);
				answer /= 2;
			}
		}
		return answer;
	}
};

/*
idea: http://articles.leetcode.com/2011/01/find-k-th-smallest-element-in-union-of.html
complexity: O(logN)
*/

int main(void)
{
	vector<int> nums1;
	nums1.push_back(1);
	vector<int> nums2;
	nums2.push_back(2);
	Solution engine;
	cout << engine.findMedianSortedArrays(nums1, nums2) << '\n';
	return 0;
}

File: 4 Median of Two Sorted Arrays/4 LL.py
class Solution(object):
	@staticmethod
	def findKth(nums1, nums2, k):
		if len(nums1) == 0:
			return nums2[k - 1]
		if len(nums2) == 0:
			return nums1[k - 1]
		if k == 1:
			return min(nums1[0], nums2[0])
		pos1 = min(k >> 1, len(nums1) - 1)
		pos2 = min(max(0, k - pos1 - 2), len(nums2) - 1)
		pos1 = k - pos2 - 2
		if nums1[pos1] <= nums2[pos2]:
			if pos1 == len(nums1) - 1 or nums2[pos2] <= nums1[pos1 + 1]:
				return nums2[pos2]
			return Solution.findKth(nums1[pos1 + 1:], nums2, k - pos1 - 1)
		else:
			if pos2 == len(nums2) - 1 or nums1[pos1] <= nums2[pos2 + 1]:
				return nums1[pos1]
			return Solution.findKth(nums1, nums2[pos2 + 1:], k - pos2 - 1)

	def findMedianSortedArrays(self, nums1, nums2):
		answer = float(0)
		total = len(nums1) + len(nums2)
		if total != 0:
			if total & 0x1:
				answer += Solution.findKth(nums1, nums2, (total + 1) >> 1)
			else:
				answer += Solution.findKth(nums1, nums2, total >> 1)
				answer += Solution.findKth(nums1, nums2, (total >> 1) + 1)
				answer /= 2
		return answer

engine = Solution()
print engine.findMedianSortedArrays([1, 2], [0, 1])